### DISCUSSION TITLE
Default list values

### DISCUSSION THREAD

**Discussion by verytable** [cuelang/cue#914] [Q&A]
**General question.** Do lists work well with defaults? If they don't I think this should be explicitly mentioned in docs.

My doubts are based on the following quote from docs:
> Lists can be thought of as structs:
>
>List: *null | {
>    Elem: _
>    Tail: List
>}

that shows that List has default (*) in it's definition. Also there are no list examples in the "Default values " section of the docs.

**Example.**
Evaluation of the following code (with cue 0.3.2)
`{ arr: [...string] | [] } & {arr: *[] | [...string]}`
results in
`arr: incomplete value [] | []`

However, this
`{ arr: [...string]} & {arr: *[] | [...string]}`
results in
`
{
    "arr": []
}
`
The first example also works with cue 0.2.2.

What is the difference between these two snippets?

**Comment by myitcv** [comment]
@verytable sorry for the slow response here. 

Lists and structs have a default value of the empty list or struct respectively:

```
exec cue export
-- x.cue --
package x


l: [...string]
s: {...}
```

gives:

```
{
    "l": [],
    "s": {}
}
```

I think the quote from the spec "Lists can be thought of as structs" is meant to illustrate another way of thinking about lists, but I can see how it is misleading in light of the above.

The behaviour of your example:

```
{ arr: [...string] | [] } & {arr: *[] | [...string]}
```

has changed with respect to `v0.2.2`. I will, however, defer to @mpvl on the finer points of that, and how it will change in the future (it relates to elimination of disjuncts, something that is a function of how disjuncts are considered equivalent or not, for some definition of equivalent).

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/914.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
